pieces.cpp

```diff
@@ -1,263 +1,158 @@
-#include <algorithm>
 #include "pieces.h"
 
-using namespace std;
-
-
-Square::Square() {
-    pos = 205;
-}
-
-void Square::rotate() {
-    // rotating results in the same shape
-    return;
-}
-
-void Square::down() {
-    if (pos >= 0 || (*board)[pos - 10] || (*board)[pos - 9]) {
-        set();
-        return;
-    }
-    pos -= 10;
-    // move up one row if too low
-}
-
-void Square::left() {
-    if (pos % 10 > 0)
-        pos -= 1;
-}
-
-void Square::right() {
-    if (pos % 10 < 8)
-        pos += 1;
+// Piece
+
+const std::vector<int>* Piece::relRowPositions = nullptr;
+const std::vector<int>* Piece::relColPositions = nullptr;
+
+Piece::Piece(Board& b) : board(b) {
+	this->color = COLOR::COLOR_EMPTY;
+	this->positionRow = PIECES::INITIAL_ABS_POSITION_Y;
+	this->positionCol = PIECES::INITIAL_ABS_POSITION_X;
+	this->rotation = PIECES::INITIAL_ROTATION;
+}
+
+// true if piece collides
+inline bool Piece::checkCollidesAtOffset(int rowOffset, int colOffset) {
+	return board.checkPositionOccupied((this->positionRow + (*relRowPositions)[rotation * PIECES::STATES_OF_ROTATION + 0] + rowOffset), (this->positionCol + (*relColPositions)[rotation * PIECES::STATES_OF_ROTATION + 0] + colOffset))
+		|| board.checkPositionOccupied((this->positionRow + (*relRowPositions)[rotation * PIECES::STATES_OF_ROTATION + 1] + rowOffset), (this->positionCol + (*relColPositions)[rotation * PIECES::STATES_OF_ROTATION + 1] + colOffset))
+		|| board.checkPositionOccupied((this->positionRow + (*relRowPositions)[rotation * PIECES::STATES_OF_ROTATION + 2] + rowOffset), (this->positionCol + (*relColPositions)[rotation * PIECES::STATES_OF_ROTATION + 2] + colOffset))
+		|| board.checkPositionOccupied((this->positionRow + (*relRowPositions)[rotation * PIECES::STATES_OF_ROTATION + 3] + rowOffset), (this->positionCol + (*relColPositions)[rotation * PIECES::STATES_OF_ROTATION + 3] + colOffset));
+}
+
+bool Piece::moveDown() {
+	bool canMove = !checkCollidesAtOffset(PIECES::MOVE_DOWN, 0);
+	this->positionRow += canMove * PIECES::MOVE_DOWN;
+	return canMove;
+}
+
+bool Piece::moveLeft() {
+	bool canMove = !checkCollidesAtOffset(0, PIECES::MOVE_LEFT);
+	this->positionCol += canMove * PIECES::MOVE_LEFT;
+	return canMove;
+}
+
+bool Piece::moveRight() {
+	bool canMove = !checkCollidesAtOffset(0, PIECES::MOVE_RIGHT);
+	this->positionCol += canMove * PIECES::MOVE_RIGHT;
+	return canMove;
+}
+
+void Piece::drop() {
+	while (moveDown());
+}
+
+bool Piece::rotate() {
```

Comment:
still not really sure what all the ifs are for in this rotate function so will probably have to explain to me or write a comment.